### PR TITLE
VISSL collate function/transforms restructure

### LINKED ISSUE #783: Standardise `ImageEmbedder` to take a `training_strategy_kwargs` argument instead of just varargs.
## 🐛 Bug

Refactor for consistency with the learn2learn integration.

### PR DESCRIPTION
**Pull request by ananyahjha93**
## What does this PR do?

Fixes #783, #782, #781 .

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? [not needed for typos/docs]
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: flash_examples/image_embedder.py [hunk 1]
```diff
@@ -11,31 +11,15 @@
...
-
-# 1. Download the data and pre-process the data
-transform = IMAGE_EMBEDDER_TRANSFORMS.get("simclr_transform")()
-
-to_tensor_transform = ApplyToKeys(
-    DefaultDataKeys.INPUT,
-    transform,
-)
 
+# 1. Download the data and prepare the datamodule
```
Comment (tchaton) [inline flash_examples/image_embedder.py:20]: Can we add some tests ? Lot of code isn't covered.

File: flash_examples/image_embedder.py [hunk 2]
```diff
@@ -11,31 +11,15 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-import torch
```
Comment (tchaton) [inline flash_examples/image_embedder.py:14]: I would create 2 examples. 1 for ImageEmbedder, and 1 for ImageEmbedder + SSL within integrations under vissl.

Comment (ananyahjha93) [reply flash_examples/image_embedder.py:14]: @tchaton for this PR, I'll keep it the current way, I can add a separate issue and PR to load vissl weights with ImageEmbedder for inference, like I did with the pretrained weights for ImageClassifier

File: flash/image/embedding/model.py [hunk 3]
```diff
@@ -57,31 +62,41 @@ class ImageEmbedder(AdapterTask):
...
+            training_strategy_kwargs = {}
+
+        backbone, _ = self.backbones.get(backbone)(pretrained=pretrained, **backbone_kwargs)
 
         metadata = self.training_strategies.get(training_strategy, with_metadata=True)
-        loss_fn, head, hooks = metadata["fn"](**kwargs)
+        loss_fn, head, hooks = metadata["fn"](head=head, **training_strategy_kwargs)
+
+        print(head)
+        exit(-1)
```
Comment (SeanNaren) [inline flash/image/embedding/model.py:99]: debug stuff?

File: flash/image/embedding/model.py [hunk 4]
```diff
@@ -100,6 +112,15 @@ def __init__(
             learning_rate=learning_rate,
         )
 
+        transform, collate_fn = self.transforms.get(pretraining_transform)(**pretraining_transform_kwargs)
+        to_tensor_transform = ApplyToKeys(
+            DefaultDataKeys.INPUT,
+            transform,
+        )
+
+        self.adapter.set_state(CollateFn(collate_fn))
```
Comment (ethanwharris) [inline flash/image/embedding/model.py:121]: Nice :smiley:

File: flash_examples/image_embedder.py (hunk 1 above)
Comment (ananyahjha93) [reply flash_examples/image_embedder.py:20]: added

File: flash/image/embedding/model.py [hunk 5]
```diff
@@ -100,6 +112,15 @@ def __init__(
...
+        self.adapter.set_state(ToTensorTransform(to_tensor_transform))
```
Comment (ethanwharris) [inline flash/image/embedding/model.py:122]: Not sure about this. I think it means that the `pre_tensor_transform` and `post_tensor_transform` etc. provided either by the user or our default transforms will still be running. So you’ll get resize, flip, etc. then the multicrop transform or whatever. Maybe we should add a state thing to allow the model to override the whole transform recipe? How should the VISSL stuff behave with users custom transforms?

Comment (ananyahjha93) [reply flash/image/embedding/model.py:122]: yeah, ideally the user should not put in their own custom transforms here, so I will add a line to override everything within the model

File: flash/image/embedding/model.py [hunk 6]
```diff
@@ -57,31 +62,38 @@ class ImageEmbedder(AdapterTask):
...
+    transforms: FlashRegistry = IMAGE_EMBEDDER_TRANSFORMS
 
     required_extras: str = "image"
 
     def __init__(
         self,
         training_strategy: str,
-        embedding_dim: int = 128,
+        head: str,
+        pretraining_transform: str,
```
Comment (ethanwharris) [inline flash/image/embedding/model.py:73]: Can this be inferred from the training strategy name? Or can you mix and match the transform recipes?

Comment (ananyahjha93) [reply flash/image/embedding/model.py:73]: you can mix and match, and even customize these recipes, we can add the logic for inferring these in the next PR

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:

**Review by SeanNaren — APPROVED** [review]
great work @ananyahjha93 looking forward to the docs :)

**Pull request merged**